Publish Safe Module exploit analysis (#87)
commonprefix/website: blog/4m-usd-safe-drain-analysis.md

---

> title: "$4M Safe Module Drain: What Happened"

Reviewer: ```suggestion
title: "$4M `SquidRouterModule` Exploit: What Exactly Happened"
```

Author: I wanted to avoid placing `SquidRouterModule` to the front plan to avoid linking it to Squid. we do mention it in the text but it might give wrong impression from the headline.

Reviewer: But which "Safe Module" got drained? It's a bit ambiguous, no?

date: "26/05/2026"

> desc: "Analysis of the third-party SquidRouterModule Safe module exploit on 25th May 2026."

Reviewer: ```suggestion
desc: "Analysis of the third-party "SquidRouterModule" Safe module exploit on May 25, 2026."
```

authors: "nikolas,ristic"
index: true
---

> **All Axelar users, user funds, approvals, and integrations are unaffected, and no action is needed. Our integration partner Squid was also not affected.**

Reviewer: ```suggestion
**All Axelar user funds, approvals, and integrations are unaffected, and no action is needed. Our integration partner Squid was also not affected.**
```

## TL;DR

- In May 2026, an exploit targeted `SquidRouterModule`, a third-party Safe module deployed at the same address on [Ethereum](https://etherscan.io/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca), [Base](https://basescan.org/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca), and [Arbitrum](https://arbiscan.io/address/0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca): `0x1f1d37a3Bf840e35c6a860c7C2dA71Fe555123ca`.

> - The observed victims were individual Safe smart accounts, not a single protocol treasury. They had enabled `SquidRouterModule` and granted delegate permissions through a related permissions system. The affected component was the `SquidRouterModule` application module, not Axelar or Squid.

Reviewer: ```suggestion
- The observed victims were individual Safe smart accounts that had enabled `SquidRouterModule` and granted delegate permissions through a related permissions system. The affected component was the `SquidRouterModule` application module, not Axelar or Squid.
```

Reviewer: What's the "related permissions system"?

Author: made it clearer

> - Across the indexed attack set, the observed drain total is approximately $3.98M across 313 attack transactions, including 308 successful drains and 5 reverted attempts, and 88 distinct victim Safes, using the investigation snapshot from 26th of May 2026.

Reviewer: I think there were way more than 5 reverted attempts.

Reviewer: I think I saw more like 1000

Author: any data that you can share? the data I indexed doesn't have that many. we can avoid listing the reverted tx number until certain.

Reviewer: I remember seeing on Etherscan a looooot of on-chain reverts. We can just say "numerous" here if we don't have exact numbers.

> - Across the indexed attack set, the observed drain total is approximately $3.98M across 313 attack transactions, including 308 successful drains and 5 reverted attempts, and 88 distinct victim Safes, using the investigation snapshot from 26th of May 2026.

Reviewer: ```suggestion
- Across the indexed attack transactions, the observed drain total is approximately $3.98M across 313 attack transactions, including 308 successful drains and 5 reverted attempts, and 88 distinct victim Safes, using the investigation snapshot from 26th of May 2026.
```

- The core issue was that `SquidRouterModule` decoded a `delegate` address from attacker-controlled cross-chain payload bytes and then used that address for local Safe permission checks.

> - After the drains, funds were consolidated across Ethereum, Base, and Arbitrum via Relay and ultimately sent as DAI to an Ethereum address in the attacker cluster.

Reviewer: Just Relay? Also they used some DEXs, right?
Do we mention somewhere that the first observed attacker account in the chain of accounts was funded via Tornado Cash? How much ETH was it funded with?

Author: other DEXes were used, I added without mentioning specifically. let me know if that's ok.

Reviewer: sounds reasonable

## Incident Summary

> On May 25th 2026 around 90 different Safe wallets were drained totalling to around $4M USD worth of tokens stolen. The exploit targeted a Safe module named `SquidRouterModule`. This module was written, deployed and operated by a third-party. `SquidRouterModule` was part of a larger Safe module suite whose deployer also deployed sibling modules for Aave, Morpho, Yearn, Uniswap V2/V3/V4, and Krystal.

Reviewer: Do we not have the exact number of Safes now? And the exact number of USD under some price assumptions that we can mention?

Author: we do, mentioned the numbers precisely.

> The identity of the deployer has not been confirmed, but given a recent [blockscan chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the party behind these contracts.

Reviewer: ```suggestion
The identity of the "SquidRouterModule" deployer has not been confirmed, but given a recent [Blockscan Chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the party behind these contracts.
```

### Step 2. The vulnerable handler trusted a delegate from payload bytes

The vulnerable destination-chain handler decoded the `delegate` value from the cross-chain payload:

```solidity
(address module, address safe, address delegate, ActionsExecutionParams memory params)
    = abi.decode(payload, (address, address, address, ActionsExecutionParams));

require(module == address(this), ...);
bridgedToken.safeTransfer(safe, bridgedTokenAmount);
_handleActions(safe, delegate, params);
```

The `module == address(this)` check only verified that the payload claimed to be intended for `SquidRouterModule`. It's later checked that the delegate value had the right permissions:

```solidity
function _handleUniV3SwapExactIn(
    address safe,
    address delegate,
    bytes memory encodedData
) internal returns (bool result) {
    _checkPermission(safe, delegate, Permissions.SWAP_PERMISSION);

    // do the swap
}
```

The core issue was therefore simple: the attacker could choose the delegate address inside the payload. If that delegate already had `"*"` permission for the victim Safe, `SquidRouterModule` accepted the request and executed the requested actions.

> The observed attacks used Axelar express execution. That was the delivery path, not the root cause. The vulnerable trust decision was inside `SquidRouterModule`'s payload handling: the destination contract treated attacker-controlled message fields as authority for a local Safe operation.

Reviewer: ```suggestion
The observed attacks used Axelar Express Execution. That was the delivery path, not the root cause. The vulnerable trust decision was inside `SquidRouterModule`'s payload handling: the destination contract treated attacker-controlled message fields as authority for a local Safe operation.
```

Reviewer: What does "delivery path" mean?

Author: made it clearer

> Some attacks called `SquidRouterModule` directly. Others went through attacker contracts that simplified the attack call. For example, the Ethereum attacker contract [`0xfac7459683cdb9b6f367b42eedfebd745dc8760c`](https://etherscan.io/address/0xfac7459683cdb9b6f367b42eedfebd745dc8760c) decompiled to a two-argument entry point taking a victim Safe and delegate. In this write-up we refer to that behavior as `drain(safe, delegate)`, but the name is not a verified ABI name.

Reviewer: I don't think "decompiled" is the right term here.

Author: I used [Lutetia](https://github.com/jose-compu/lutetia) but similar tools exist, all termed as decompilers

Reviewer: Isn't it just the keccak of the signature? You can find the function name and params by using a DB like the one `cast 4d` uses, no? What's the decompilation part?

Author: I decompiled the bytecode to find out not just the signature but the arguments and the code, and used AI to prettify it. The selector hash doesn't reveal the signature of the function, the attacker surely wouldn't use a known signature available in 4byte (I checked).

Reviewer: Gotcha

    UNI_V3_SWAP_EXACT_OUT,
    ERC20_APPROVE,
    PERMIT2_APPROVE,
    NATIVE_WRAP,
    NATIVE_UNWRAP
}
```

Observed attacker wallets include:

- Attacker wallet 1: [`0x7c82cb4b2909c50c7c0f2b696eee7565e0a23bb8`](https://etherscan.io/address/0x7c82cb4b2909c50c7c0f2b696eee7565e0a23bb8), the main operator wallet. It deployed attacker contracts, executed most attacker-contract calls, used Relay during consolidation, and sent DAI to the final Ethereum wallet.
- Attacker wallet 2: [`0x9bdc730183821b6bb2b51be30b77c964fa645b91`](https://etherscan.io/address/0x9bdc730183821b6bb2b51be30b77c964fa645b91), a co-operator wallet. It deployed attacker contracts and funded attacker wallet 1 with operational gas across chains.
- Attacker wallet 3: [`0x7e54c729148a95bca651f3214ac9ebefd3fb1271`](https://etherscan.io/address/0x7e54c729148a95bca651f3214ac9ebefd3fb1271), which deployed fake tokens and rigged Uniswap V3 pools and also made direct calls to `SquidRouterModule`.
- Attacker wallet 4: [`0xc8ef4003d9db3863b9af26afcf2275378bfa83e4`](https://etherscan.io/address/0xc8ef4003d9db3863b9af26afcf2275378bfa83e4), a secondary operator wallet that made direct calls to `SquidRouterModule`.

> - Final Ethereum wallet: [`0xa447f71782135ab96a71374271a749ff7aa54859`](https://etherscan.io/address/0xa447f71782135ab96a71374271a749ff7aa54859), which received 3,070,767.09 DAI from attacker wallet 1.

Reviewer: ```suggestion
- Attacker wallet 5: [`0xa447f71782135ab96a71374271a749ff7aa54859`](https://etherscan.io/address/0xa447f71782135ab96a71374271a749ff7aa54859), which received the consolidated 3,070,767.09 DAI from attacker wallet 1.
```

`SquidRouterModule` did not expose a generic transfer action or arbitrary Safe call. Its action menu was limited to swaps, approvals, and native wrap/unwrap operations, and approval spenders were constrained to Squid Router, supported Universal Routers, or Permit2. The attacker therefore could not simply approve an attacker wallet or transfer tokens directly out of the Safe.

The extraction path was the swap itself. By routing through attacker-created Uniswap V3 pools paired against worthless attacker-issued tokens, the Safe sent real assets into the pool and received those worthless tokens back. The attacker, as the pool LP, could then withdraw the real assets from the pool.

The attacker used multiple fake tokens and created Uniswap V3 pools as drain destinations. One example swap on Ethereum is [`0x8de614bdb7acf5dcbdfe5ce8ed17ec2a2058e7708e6a4cf44f0e523c72df24c3`](https://etherscan.io/tx/0x8de614bdb7acf5dcbdfe5ce8ed17ec2a2058e7708e6a4cf44f0e523c72df24c3):

- The victim Safe sent 32,000 apyUSD.
- The Safe received a fake `"USDT"` token: `0xd7ef4cb5c57a4362a48abf4d33c5b9470d6681e1`.
- The fake token and the corresponding Uniswap V3 pool were both deployed by attacker wallet 3.

> ## What Was Affected

Reviewer: ```suggestion
## Who Was Affected
```

The affected accounts were Safe smart accounts that had enabled `SquidRouterModule` and had matching delegate permissions. The observed victims were individual Safe smart accounts, not a single protocol treasury.

Axelar and Squid were not affected. Their users, funds, approvals, and integrations are unaffected, and no action is needed.

> The observed drain total was approximately $3.98M across 313 attack transactions, including 308 successful drains and 5 reverted attempts, and 88 distinct victim Safes:

Reviewer: Can we add diagrams like this one for each chain: https://x.com/timmi_arno/status/2058914390095671504?s=20 ?
Maybe we can just use Arkham like on that tweet.

- UNI: about 2,603.6 tokens ($8,748).
- CFG: about 25,463 tokens ($7,749).
- MORPHO: about 3,009.3 tokens ($7,192).
- ENA: about 56,582 tokens ($5,718).
- FLUID: about 1,406.7 tokens ($4,220).
- USDT: about 3,533.1 tokens ($3,533).
- EUL, LBTC, ZRO, LINK, WLD, Anon, AERO, EURC, SKY, SLVon, SOL, msETH, XAUt, COOKIE, AVNT, USDT0, Cake, AUKI, and SAPIEN, together totaling about $12,171.

Base was the largest loss surface in the observed data, accounting for roughly 71% of the total. Ethereum accounted for roughly 27%, and Arbitrum accounted for roughly 2%.

Across the current dataset, 76 of the 88 observed victim Safes were drained on two or more chains.

## Fund Movement

After the drains, funds were pulled together across Ethereum, Base, and Arbitrum. The consolidation hub was attacker wallet 1 ([`0xa447f71782135ab96a71374271a749ff7aa54859`](https://etherscan.io/address/0xa447f71782135ab96a71374271a749ff7aa54859)), which consolidated funds via Relay. The funds were bridged, swapped and eventually consolidated as 3,070,767.09 DAI.

At the time of the investigation notes, the DAI had not moved out of attacker wallet.

<br/>

## FAQs

<br/>

### Were Axelar or Squid compromised?

> No. The affected component was contract called `SquidRouterModule`, a third-party application module that incorrectly used Axelar's cross-chain execution protocol. Based on a recent [blockscan chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the victim platform and the author of these contracts.

Reviewer: ```suggestion
No. The affected component was a smart contract called `SquidRouterModule`, a third-party application module that incorrectly used Axelar's cross-chain execution protocol. Based on a recent [blockscan chat message](https://etherscan.io/tx/0xa7db9793d4978a26ee01c2a6cbf5ba154925a7532e31ae731f7f08ba4295ee70) it's possible [NewMarketTrading](https://newmarkettrading.com) is the victim platform and the author of these contracts.
```
